### ISSUE TITLE
proposing a new mechanics API

### ISSUE THREAD

**Issue by radarsat1** [siconos/siconos#29]
This will take a bit of explaining, but it comes from discussions with @vacary and @bremond, and is related to the integration of Siconos with Gazebo.  It's a little early to show this, it is very far from done and not even ready to put on master branch, but I would like to hear comments from the team on the approach before I make too much progress.

For a brief explanation, please see the comments of 4eee836134c57b77c97d5a6b22270fb3ab698995 and changes of my branch [mechanics-proposed](https://github.com/radarsat1/siconos/tree/mechanics-proposed/mechanics/src/proposed)

Here is a run-down.  As discussed with @bremond, some goals are,
- Remove references to Bullet classes in user code.  Make it easier to switch collision back-ends.  No more BulletDS.
- Detach definition of bodies from their shapes.  No more SphereDS, etc.

Additionally, the API should make it easy to describe scenes and instantiate objects corresponding to user software objects such as the Link/Collision/Shape distinction in Gazebo, both in C++ and in Python.

Personally I additionally wanted to avoid any unnecessary up/down-casting, which was quite difficult to achieve with an extended hierarchy like Shape→Sphere→BulletSphere, etc.  This approach was viable, using a "factory" to generate Bullet-specific objects, but passing the pointers to the DS unavoidably required some casting.  This is not just aesthetic — I want it to be impossible at compile time to pass a Bullet object to an ODE back-end, for example.

The solution I have converged on is to treat the scene description as an abstraction, and to have the Bullet-specific code "implement" this description.  In [testContact.cpp](https://github.com/radarsat1/siconos/blob/mechanics-proposed/mechanics/src/proposed/testContact.cpp), you can see that one creates a scene,

```
SP::BodyDS body(new BodyDS());
SP::Contactor contactor(new Contactor());
SP::SiconosSphere sphere(new SiconosSphere(1,2,3,1.0));
contactor->addShape(sphere);
body->setContactor(contactor);
```

and then a particular broadphase implementation takes this scene as input:

```
SP::BulletBroadphase broadphase(new BulletBroadphase());
broadphase->buildGraph(contactor);
```

Here, `SP::BulletBroadphase` could be switched for another implementation of the `SP::SiconosBroadphase` interface.

The visitor framework turned out to be very useful for this, since the appropriate SiconosSphere or SiconosBox visitor is called, alleviating any need for casting.

Behind the scenes, `buildGraph()` traverses the scene (it will need to take a different argument eventually, probably the Model), and produces a set of `btCollisionObject`/`btCollisionShape` objects that reproduce the scene.  This is then used in,

```
broadphase->performBroadphase();
```

which should produce indexSet1.

So, the only complicated part is that the scene may be updated; for instance, one might change the radius of a `SiconosSphere` after `buildGraph()` has been called.  How to maintain these updates?

My solution thus far — and I'm definitely open to ideas here — is to have each shape contain a handler that is called any time an object's properties are changed.  This handler is installed by the broadphase implementation during `buildGraph()`, and is an object that may contain extra broadphase-specific code and data.  For now, it simply puts the objects on a "dirty" list, but it can be used to update the associated `btSphereShape` for example.  Or, a single handler could be used for all objects that contains a hash table associating Bullet and Siconos objects.  Different approaches are possible.

The important thing is that all Bullet-specific code is very well separated, and could be easily switched for a different collision engine.  Meanwhile, the `SiconosSphere`/`SiconosBox` etc classes can be expanded to include a set of object types we desire to support — this could be based on SDF / URDF for example, but could also perhaps include 2D objects for a different back-end.  The back-end is free to not support certain objects, and will only fail if it encounters an object it does not implement.

Things I'm not sure of:
- whether the `Contactor` class is really needed, it only maintains a list of shapes.
- whether joints need to be considered here.  Perhaps they are just Relations between BodyDS and can be ignored with regards to the collision engine interface.
- whether `Contactor`/`SiconosShape` instance can be shared between many DSs.  This could be useful for particle models, but at the moment the handler is assumed to be instantiated for each object.  Alternatively it could be shared.

I'm interested to hear your comments.  I'll close this issue when the branch is either merged (not necessarily finished!) or decided not to merge.

**Comment by radarsat1** [comment]
An update on the current status of this new mechanics API. Currently we can instantiate sphere, box, and plane shapes.  I spent quite some time trying to figure out how to tweak the Bullet objects (btBoxShape etc) to work in an expected way, but for now I have settled on simply using btConvexHullShape for all objects.  (Ensuring that the same collision algorithm is used on each.)  However, the choice of which "implementation" of these is used can be changed with some ifdefs at the top of `BulletBroadphase.cpp`.  Importantly, one can instantiate these shapes (almost) without caring about which implementation is used.

I say almost, because of the different performance under different algorithms -- after quite some work it was almost impossible to find a consistently "best" configuration for which shapes to use and how to set the margins.  Therefore I resigned to exposing the concepts of "inside" and "outside" margins, which are considered properties of any shape, although used somewhat differently between btBox/SphereShape and btConvexHull.  The "outside" margin is removed during distance measurement in `Relation::computeh()`.  The selection of margins must take into account:

- the size of the object
- how fast it will move (displacement per time step affects the probability that it will cause a GJK overlap)
- how much rounding of the corners is acceptable (an effect of the "inside" margin)

Unfortunately it seems that simply setting a very large "outside" margin doesn't always work.  Also, the outside margin may have an effect on how you can initialize the objects.

It may also affect initialisation.  When objects are too closed together at initialisation, such that their outside margins overlap, they will start with existing contact points.  This may or may not work well, still to be determined.  Just something to be aware of.

----

What is implemented currently,

- Box, Sphere, Plane
- Global options including:
  * contact-breaking threshold (Bullet global `gContactBreakingThreshold`)
  * world scaling -- positions and dimensions are scaled by a factor, may help with implementing small or large objects.
- Python interface, demonstrated in [contact_test.py](https://github.com/radarsat1/siconos/blob/mechanics-proposed/mechanics/src/proposed/test/contact_test.py)
- Mechanics IO interface, demonstrated in [ProposedIO/cube.py](https://github.com/radarsat1/siconos/blob/mechanics-proposed/examples/Mechanics/ContactDetection/ProposedIO/cube.py)
- The Contactor class turned out to be useful for putting together one or more shapes as static objects.  Static objects are simply contactors that are added directly to the collision engine, but are not assigned to a Body.

-----

What needs work,

- Add support for convex hull objects (includes computing shrunk hull)
- Decide on how to name this thing
  * I don't think it should be called SiconosBroadphase, more like SiconosCollisionWorld or something
  * Should "Siconos" be prepended everywhere? Otherwise we may need a namespace, as names like Box, Sphere, etc., will clash with users (e.g. Gazebo)
  * BodyDS, BodyTimeStepping, etc.. ok?
  * Rename "proposed" folder.  Port existing "bullet" examples.
- Add support for existing alternatives,
  * Non-bullet objects, like Circle, Sphere, Disk from "bodies" folder
  * Port `SpaceFilter.cpp` to be used through the SiconosBroadphase interface
  * OCC and Mechanisms stuff (seems quite involved..)
  * Joints (maybe nothing to do)
- A few missing features
  * Shape offsets (currently all shapes are assumed centered at the Body origin)
  * Compound objects

Ultimately the idea is to unify the existing Mechanics interfaces and examples such that user code can describe a scene without reference to back-end specific interfaces such as Bullet.

**Comment by radarsat1** [comment]
Some other things to do,

- possible to handle updates to shapes such that shape / contactors can be shared between bodies?

- review inefficiencies in the current way that shape properties are updated.  Items are put on a dirty list and the collision shape is completely updated, even when just the position changes.  This item is likely tied to the above issue as well as "shape offsets".

**Comment by radarsat1** [comment]
Forgot an important feature!

- Add control over non-smooth law, e.g. re-implement the existing "collision group" idea, or some idea of object materials and collision response pairs.

**Comment by radarsat1** [comment]
I have performed a big refactor/cleanup of the branch now and here I
will describe it so we can do a proper code review:

Summary of changes:

* Add `InteractionManager` to kernel -- called by `TimeStepping`
  during `advanceToEvent` and does two things: requests a visitor from
  the `InteractionManager` to pass over all DSs.  This gives an
  opportunity to update DS-related data (here, collision objects);
  secondly, calls `updateInteractions` which creates or destroys
  interactions before calling the solver.  This removes the need to
  derive from `TimeStepping`, a "manager" is instead something
  inserted into to the simulation to handle dynamic relations between
  DSs.  I felt this could go in the kernel because it could be used in
  other ways for e.g. electrical or biological simulations that
  require managing large but sparse numbers of interactions.

* Get rid of unnecessary maps/lists.
  - shape->collision and ds->shape, etc maps are replaced by a single
    list of "records" for each shape type mapping ds->(shape,collision
    shape,collision object,offset)
  - this record is pointed at by the `btCollisionObject::userDataPointer`
  - so now the handler has enough information to compute the new
    shape position and modifying the `btCollisionObject` transform
    directly without storing it somewhere.
  - visitors over contactors replaced by `createCollisionObjectsForBody`
  - "dirty" shape lists

* Shapes no longer have the "onChanged" handler: for shape sharing it
  is important not to store data or callbacks in the `SiconosShape`.
  Instead, whenever the shape properties are changed, a version
  counter is incremented. The `InteractionManager` then has the
  opportunity to notice the change in version number and update the
  collision shape to correspond.

* The above change means that now there is a separation between shape
  properties updates and position updates, so it should be more
  efficient when shapes do not change. (Previously shape properties
  were changed on every time step even if they didn't change!)

* `SiconosBroadphase` renamed `SiconosCollisionWorld`
* `BulletBroadphase` renamed `SiconosBulletWorld` (for namespacing in user code)
* The base class only adds a place to store static contactors and
  non-smooth laws. It does not keep a list of BodyDSs or handle the
  simulation in any way.
* The simulation is maintained by the user as any Siconos simulation,
  the SiconosCollisionWorld does not track DSs!  That job is left to
  `NonSmoothDynamicalSystem` as in any Siconos simulation.  The only
  difference for a mechanics simulation is if `BodyDS`s are found in
  the NSDS, then the `SiconosCollisionWorld` will visit them.

* Much more code has been moved to the hidden PIMPL object (`SiconosBulletWorld_impl`)

Still to do,

* Re-add support for dynamic shape changes using version counter.

* Add support for existing alternatives,

  - Non-bullet objects, like `Circle`, `Sphere`, `Disk` from "bodies" folder
  - Port `SpaceFilter.cpp` to be used through the `SiconosBroadphase` interface
  - OCC and Mechanisms stuff (seems quite involved..)
  - Joints (maybe nothing to do)

* Decide on whether the "static contactor queue" is required --
  explanation: I noticed that changing the order of adding the
  collision objects changed the collisions that we get, which chagned
  the simulation.  Previously the initial objects were added and then
  the static contactors, but after some changes this order was changed
  and my simulation outcome changed.  I added a queue to ensure static
  contactors were added *after* the initial dynamic objects are
  imported.  However, perhaps this is not necessary.

**Comment by radarsat1** [comment]
On further thought I decided to rename `SiconosMechanicsWorld` to `SiconosCollisionManager`, and `SiconosBulletWorld` to `SiconosBulletCollisionManager`.  More descriptive, but maybe more boring.. ;)

**Comment by vacary** [comment]
Meeting @bremond @radarsat1 and myself

- InteractionManager
    -   factorize group info and insertNonSmoothLaw

**Comment by radarsat1** [comment]
Progress report:
* SiconosContactor -- introduced SiconosContactorSet, and SiconosContactor now remains more like the role of "Contactor" in mechanics_io.  It maintains a single offset and shape, and SiconosContactorSet is a vector of these.  In addition, I introduced SiconosContactorBase used internally to abstract the offset of the whole SiconosContactorSet.  It can be specified directly for static contactors (allowing to re-used SiconosContactorSets at different locations) or take the form of a BodyDS for moving objects.
  - still playing with this.. I might drop it and just use an SP::SiconosVector, as it introduces multiple inheritance for BodyDS which may not be the best idea and turned out to be a bit less elegant that I thought anyway.
* BodyDS -> i added some inertia things, but more to NewtonEulerDS, which BodyDS inherits from.  I agree that contactor-related inertia computations could go in BodyDS or SiconosContactor!
* I have the impression that the restart problem is gone after my last changes, still testing.

**Comment by radarsat1** [comment]
Result from some overnight tests are that I am able to resume the simulation, but there appears to be problems with contact with some static objects.  Also, small convex hulls tend to tunnel through the floor sometimes. I think it may be related to the order in which collision objects are added to the scene, and how this affects the contact pairs that are generated, (and the normals) but it's not clear to me.

(I've noticed that the scene behaviour changes quite drastically depending on what order collision objects are created.)

**Comment by radarsat1** [comment]
An example of what I mean.  There are 3 walls, and none of the objects made contact with them.  But they did make contact with the floor!  But still managed to fall through the floor sometimes..

![vview screenshot](http://i.imgur.com/kYosqHF.png)

I've noticed some of this tunneling behaviour in some less complicated simulations as well, it seems Bullet does not always make the same contact points.

**Comment by vacary** [comment]
Is it only when you restart a computation or also for simulations from scratch ?
Do you think that it is a feature of the new mechanics-proposed interface or it was the same before ?

**Comment by radarsat1** [comment]
I think it's related to the order of when btCollisionObjects are added to the simulation.  Actually, it is the same kind of problems we were having long ago with contacts not being always predictable.  I am finding that you can usually find some set of margin parameters and gContactBreakingThreshold that more or less solves it but it is annoying.

The reason it is more an issue now with time_of_birth and restarting the simulation is that the order in which dynamic and static collision objects are added is not preserved.  I am still looking into it, I hope it find some kind of solution.  At least one implication is that sometimes contact points are created where the first object is the static one (with no DS) and therefore the contact normal is in the wrong direction.  I handled this by flipping the normal in this case, and it works but there does seem to be more problems with simulations when the normal is flipped, and I am not sure why.

That might be a red herring however, because it seems that the biggest problems come from there being NO or not enough contact points generated in some cases.  As an example, I added a #define at the top of SiconosBulletCollisionManager.cpp called QUEUE_STATIC_CONTACTORS.

The effect of defining it is that when loading a static object, the btCollisionObject is not created immediately but instead held in a queue and added at the beginning of the first time step, *after* the dynamic objects have been added.  It has quite an effect!  See this animation, the _only_ difference is the QUEUE_CONTACT_POINTS defined or not defined:

![animation showing QUEUE_CONTACT_POINTS](https://j.gifs.com/nZY77l.gif)

Unfortunately this is not "a solution" since I can easily find more examples where it falls through in both cases.  Just play with the NonSmoothLaw, margins, and breaking threshold parameters and you can easily find bad configurations.  So I am still trying to understand..

What I will do next is try to ensure that the btCollisionObject creation is in _exactly_ the same order as on the master branch, and that all parameters are the same (margins etc), and see if that at least makes it easier to compare.

**Comment by radarsat1** [comment]
I should mention that I've now addressed all points except:

* computation of mass matrix and inertia with Huygens theorem (no inertia matrix assigned to shape, only to DS.. ideas welcome)
* native, 2D, mechanisms, OCC

**Comment by radarsat1** [comment]
In example of the tetrahedron falling that I posted above, if I set the insideMargin to 0.04 (which is the same as the default for the original mechanics IO) then I get the same (correct) behaviour in both initialization orders.  Therefore it seems to me that perhaps the initialization order, although it affects the behaviour, is not as important as simply getting the margins exactly right for any given shape. I'll continue testing tomorrow.

**Comment by radarsat1** [comment]
Good news, after a day of debugging I found several issues due to or not addressed in recent changes.
* contactorset positions were incorrect for static objects
* default margins were incorrect for boxes and spheres, making one of the dimensions effectively 0
* inertia was supported but not set from mechanics_io.py
* added code to update the bounding box after shape dimensions are changed (shouldn't have affected previous results)

here is the previous convex hulls example running with 100 objects, now works correctly. This was done using the 2-step simulation-resume method, so it's clear that this works now.
![convex hulls example video](http://i.imgur.com/yb016Lg.gif)

**Comment by vacary** [comment]
Well done. I have no other question (-:

**Comment by radarsat1** [comment]
I'll add that the problem with the tetrahedron falling through the floor when the static contactors are added to the scene first is still an issue.  However, it's hard to draw conclusions -- it seems to have more problems in static contactors are added first, so I leave the "queued static contactors" behaviour for now.

This would correspond with more contacts having flipped normals.  The reason this was not a problem before was because the entire scene was destroyed whenever a new object was added, and re-initialised, with the static contactors last, ensuring that they are always the second object in the collision pair.  I could revert to this behaviour.

On the other hand, it is hard to pin down because with the tetra.py example, by playing with margins and the settings in SiconosBulletCollisionManager.cpp (defining/undefining QUEUED_STATIC_CONTACTORS and USE_CONVEXHULL_FOR_BOX), it's possible to find good and bad behaviours for any of the conditions.  In practice I think it may be just a matter of making sure the margins are set correctly, and flipped normals are avoided.  (This could be made an error condition if the normal-flipping solution really doesn't work, which may be so.. but I'm not sure why.)

**Comment by radarsat1** [comment]
Well, I wanted to submit a pull request tonight but after testing all the examples today, it appears this problem with missing contact points (cf. the tetrahedron example above) is quite prevalent, and rather worse than I thought.  Even with n_cubes.py they fall through the floor, which I don't recall happening before, so now I am wondering if I really introduced a subtle problem somewhere.  Perhaps I am not iterating some contact points, or wrongly rejecting some DS pairs in some particular case.  I will have to continue to investigate.

I will also need to add Cylinder for some of the examples, and I should make a Trimesh shape as well.

In the meantime however, I have made it so that mechanics_io can select which implementation you want to use, and chooses the correct default classes accordingly.  So this means all the examples can be switched between implementations with one line at the top of the file.  I have added this line to select the original Bullet backend for examples that do not work correctly with mechanics-proposed.  Therefore everything in the examples/Mechanics/MechanicsIO directory currently "works."  (An exception may be the OCC example, which I cannot test as I still have not succeeded in compiling Siconos with OCC.)

So the two implementations can co-exist for a while until this last problem with missing contacts is worked out.  I don't know if it's a good idea to merge yet, however.

**Comment by radarsat1** [comment]
It appears the problem was mainly caused by a bad default contact breaking threshold.  All examples fixed now, very few non-default overrides needed.  A couple for the floor which leads to too thin insideMargin, so I set it to 0.04 in the examples.  (The floor is 0.5 deep, and I multiply by 0.02 to set an insideMargin of 0.01 which appears to be too thin, cubes fall through.)  Perhaps a "minimum" insideMargin is needed.. it really comes down to what is "normal" usage and how much the user should have to think about the margins. Difficult as the margins really depend both on object size and time step if I understand correctly.

**Comment by radarsat1** [comment]
Pull request #86 issued.

**Comment by radarsat1** [comment]
Added support for SiconosMesh.

**Comment by radarsat1** [comment]
![video of bunny](http://i.imgur.com/2xxAGLJ.gif)

Unfortunately Bullet doesn't provide mesh-mesh collision or this video would be more entertaining.

**Comment by vacary** [comment]
Cool. 
Are you sure for the mesh to mesh contact detection? I saw some simulation of @bremond with two meshes

**Comment by radarsat1** [comment]
I did run two bunnies and they passed through each other.  But brief research showed a lot of "btBvhTriangleMeshShape is intended for static shapes only..." which I translate to "we don't support mesh-mesh collision."  I can try again! ;)

In any case, probably better to eventually support some kind of convex decomposition.  There is one algorithm in GImpact but I'm not sure it's actually available through Bullet..

**Comment by radarsat1** [comment]
It appears that if you use btConvexTriangleMeshShape it will perform a convex collision based on the mesh vertices, is it possible that this is what @bremond used?  It will not be an exact mesh-mesh collision, but rather a convex hull that allows to share the same points in memory as the mesh used for display.

**Comment by bremond** [comment]
I can remember that I made sims with some gears (some simple ones and
also the Citroen gears) and some chains. I think I used btGimpactShape
(to be confirmed) and it (nearly) worked (nearly because I had to fight
against bullet scaling issues for the gears and my solution was probably
not correct). Collision with two bunnies were also ok. I did not kept
them in examples because I was not sure about copyright issues (I
quickly took the meshes from an online CAD site).

**Comment by radarsat1** [comment]
Now I left the bunnies because I found a creative commons attribution license without the non-commercial clause ;)  (Documented in mesh.py)

I can try with btGimpactShape.

**Comment by radarsat1** [comment]
Oh, I will note that one thing that is clearly harder to do with non-convex shapes is the "shrinking" of the hull necessary to offset the margins.  Therefore the margins only function as external margins for the mesh.

**Comment by bremond** [comment]
An old discussion about that

http://www.bulletphysics.org/Bullet/phpBB3/viewtopic.php?f=9&t=2849

I note 2 points:

 - For non-moving static world geometry, it is best to use btBvhTriangleMeshShape

 - In general, convex shapes shouldn't have over 100 vertices

For the first point, I am not really sure but maybe we can also consider
everything is static on bullet side (I do not see where the movement
infers with bullet and also what we are doing at this point has nothing todo
with bullet continuous contact detection which is involved in fast
moving object).

**Comment by radarsat1** [comment]
Indeed, in SiconosMesh a btBvhTriangleMeshShape is "moved" and it works fine! It is only that it doesn't generate contact points with another btBvhTriangleMeshShape.  Perhaps it works with GImpactShape, I will test when I have a chance.

**Comment by radarsat1** [comment]
Closed by merge #86
